ss_admin_console/main/default/classes/SSACDataFilter.cls

```diff
@@ -0,0 +1,115 @@
+                            break;
+                        }
+                    }
+                if (matchedNonKeyFields == nonKeyFields.size()) {
+                    results.add(new SSACResult()
+                        .setStatus(DiagnosticConstants.DUPLICATED_STATUS)
+                        .setRecordKey(recordKeysFingerprint)
+                        .setSObjectApiName(String.valueOf(sObjectApiName)));
+                }
+            } else {
+                results.add(new SSACResult()
+                    .setStatus(DiagnosticConstants.OK_STATUS)
+                    .setRecordKey(recordKeysFingerprint)
+                    .setSObjectApiName(String.valueOf(sObjectApiName)));
+            }
+        }
+        return results;
+    }
+
+    @TestVisible
+    private Map<String, SObject> generateExistentMap(List<SObject> existentRecords) {
+        Map<String, SObject> existentMap = new Map<String, SObject>();
+              for (SObject existSObj : existentRecords) {
+            String recordKeysFingerprint = '';
+            for (FieldReference keyField : keyFields) {
+                recordKeysFingerprint += String.valueOf(keyField.getFrom(existSObj)).toLowerCase() + '_';
+            }
+            recordKeysFingerprint = recordKeysFingerprint.removeEnd(recordKeysFingerprint.right(1));
+            existentMap.put(recordKeysFingerprint, existSObj);
+        }
+        return existentMap;
+    }
+
+    // // existent sobject extraction
+    @TestVisible
+    public List<SObject> getExistentRecords(List<SObject> records) {
+        String query = this.buildQuery(records);
+        return Database.query(query);
+    }
+
+    // // formatting map with all fields and its values
+    @TestVisible
+    public Map<String, Set<Object>> getKeyFieldValues(List<SObject> records) {
+        Map<String, Set<Object>> constantValues = new Map<String, Set<Object>>();
+        for (FieldReference field : keyFields) {
+            Set<Object> values = new Set<Object>();
+            for (SObject obj : records) {
+                values.add(field.getFrom(obj));
+            }
+            constantValues.put(String.valueOf(field), values);
+        }
+        return constantValues;
+    }
+
+    @TestVisible
+    private String buildQuery(List<SObject> records) {
+        Map<String, Set<Object>> keyFieldValues = this.getKeyFieldValues(records);
+        String result = 'SELECT Id, ';
+        if (dataFields != null) {
+            for (FieldReference field : dataFields) {
+                result += field.toString() + ',';
+            }
+        }
+        result = result.removeEnd(result.right(1));         
+        result += ' FROM '+ String.valueOf(sObjectApiName) + ' WHERE ';
+        for (FieldReference field : keyFields) {
+            result += field.toString() + ' IN (';
+            for (Object value : keyFieldValues.get(field.toString())) {
+                result += (value instanceof String) ? '\'' + value + '\',' :  value + ',';
```

Comment:
Escape quotes in string values with String.escapeSingleQuotes() method to prevent SOQL injections.